pkg/local_object_storage/engine/control.go

```diff
@@ -54,3 +69,70 @@ func (e *StorageEngine) Close() error {
 
 	return nil
 }
+
+// executes op if execution is not blocked, otherwise returns blocking error.
+//
+// Can be called concurrently with setBlockExecErr.
+func (e *StorageEngine) exec(op func() error) error {
+	e.blockExec.mtx.RLock()
+	defer e.blockExec.mtx.RUnlock()
+
+	if e.blockExec.err != nil {
+		return e.blockExec.err
+	}
+
+	return op()
+}
+
+// sets the flag of blocking execution of all data operations according to err:
+//   * err != nil, then blocks the execution. If exec wasn't blocked, calls close method.
+//   * otherwise, resumes execution. If exec was blocked, calls open method.
+//
+// Can be called concurrently with exec. In this case it waits for all executions to complete.
+func (e *StorageEngine) setBlockExecErr(err error) error {
+	e.blockExec.mtx.Lock()
+	defer e.blockExec.mtx.Unlock()
+
+	prevErr := e.blockExec.err
+
+	e.blockExec.err = err
+
+	if err == nil {
+		if prevErr != nil { // block -> ok
+			return e.open()
```

Comment:
If it was previously closed, all pools were released, and I thus I suppose shouldn't be working. Did you test this scenario?

Reply:
Good scenario, I'll add the unit test for it.

Reply:
12d8f63687bc5b36c79aa4caa21a59a613c68649